src/js/reducers/datasetsReducer.js

```diff
@@ -38,6 +39,41 @@ function datasetsReducer(state, action) {
     }));
   }
 
+  if (action.type === ACTIONS.FILTER_DATASET) {
+    state = setIn(state, action.id + '._filter',
+    Immutable.fromJS({
+      expression: action.expression
+    }));
+  }
+
+  var transforms = getIn(state, action.id + '._transforms');
+  if (action.type === ACTIONS.ADD_DATA_TRANSFORM) {
+
+    if (!transforms) {
+      transforms = [];
+    }
+    var result = deepEquality.indexOf(action.transformSpec, transforms);
+    if (result === -1) {
+      // if transform doesn't exist already
+      transforms.push(action.transformSpec);
+    }
+
+    state = setIn(state, action.id + '._transforms', transforms);
```

Comment:
Aha, part of the issues you're seeing is because `transforms` is a plain JavaScript array and then yes you're right, the reference to the array stays the same and so React never re-renders the component. Instead, as we do elsewhere in this file, you'll want to call `Immutable.fromJS` to convert a JavaScript array back to its equivalent ImmutableJS primitive _before_ putting it back in the store.